modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/coordination/rdbms/DatabaseSlotAgent.java

```diff
@@ -25,244 +29,639 @@
+            }
+        }
+        return assignedSlots;
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public TreeSet<Slot> getAllSlotsByQueueName(String nodeId, String queueName) throws AndesException {
+
+        String task = "retrieve all slots for queue: " + queueName + " and node: " + nodeId;
+
+        TreeSet<Slot> allSlotsForQueue = null;
+        for (int attemptCount = 1; attemptCount <= MAX_STORE_FAILURE_TOLERANCE_COUNT; attemptCount++) {
+
+            waitUntilStoresBecomeAvailable(task);
+            try {
+                allSlotsForQueue = andesContextStore.getAllSlotsByQueueName(queueName);
+                break;
+            } catch (AndesStoreUnavailableException e) {
+                handleFailure(attemptCount, task, e);
+            }
+        }
+        return allSlotsForQueue;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void reassignSlot(Slot slotToBeReassigned) throws AndesException {
+
+        String task = "delete slot assignment with start message id: " + slotToBeReassigned.getStartMessageId()
+                      + " and end message id: " + slotToBeReassigned.getEndMessageId();
+
+        for (int attemptCount = 1; attemptCount <= MAX_STORE_FAILURE_TOLERANCE_COUNT; attemptCount++) {
+
+            waitUntilStoresBecomeAvailable(task);
+            try {
+                andesContextStore.deleteSlotAssignment(slotToBeReassigned.getStartMessageId(),
+                        slotToBeReassigned.getEndMessageId());
+                break;
+            } catch (AndesStoreUnavailableException e) {
+                handleFailure(attemptCount, task, e);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void deleteOverlappedSlots(String nodeId) throws AndesException {
+        //Not necessary in RDBMS mode, because RDBMS mode uses single table to store information
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void updateOverlappedSlots(String nodeId, String queueName, TreeSet<Slot> overlappedSlots)
+            throws AndesException {
+
+        //We don't need to check whether the store is available here since it is handled in the lower level method
+        for (Slot slot : overlappedSlots) {
+            this.setSlotState(slot.getStartMessageId(), slot.getEndMessageId(), SlotState.OVERLAPPED);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<String> getAllQueues() throws AndesException {
+
+        String task = "retrieve all queue";
+
+        Set<String> allQueues = null;
+        for (int attemptCount = 1; attemptCount <= MAX_STORE_FAILURE_TOLERANCE_COUNT; attemptCount++) {
+
+            waitUntilStoresBecomeAvailable(task);
+            try {
+                allQueues = andesContextStore.getAllQueues();
+                break;
+            } catch (AndesStoreUnavailableException e) {
+                handleFailure(attemptCount, task, e);
```

Comment:
We can assign allQueues to an empty set here. Otherwise this method will return a null object when an exception occurs which can cause null pointers if not specifically checked for null in every place it is used.